extensions/grpc_logging/plugin.cc

```diff
@@ -0,0 +1,182 @@
+#include "google/protobuf/util/time_util.h"
+
+using google::protobuf::util::JsonParseOptions;
+using google::protobuf::util::Status;
+using istio_ecosystem::wasm_extensions::grpc_logging::PluginConfig;
+using istio_ecosystem::wasm_extensions::grpc_logging::WriteLogRequest;
+
+static RegisterContextFactory register_gRPCLogging(
+    CONTEXT_FACTORY(PluginContext), ROOT_FACTORY(PluginRootContext));
+
+PluginRootContext::PluginRootContext(uint32_t id, std::string_view root_id)
+    : RootContext(id, root_id) {
+  cur_log_req_ = std::make_unique<WriteLogRequest>();
+  log_entry_count_ = 0;
+}
+
+bool PluginRootContext::onConfigure(size_t configuration_size) {
+  // Parse configuration JSON string.
+  std::string configuration = "{}";
+  if (configuration_size > 0) {
+    auto configuration_data = getBufferBytes(
+        WasmBufferType::PluginConfiguration, 0, configuration_size);
+    configuration = configuration_data->toString();
+  }
+
+  JsonParseOptions json_options;
+  json_options.ignore_unknown_fields = true;
+  PluginConfig config;
+  Status status = JsonStringToMessage(configuration, &config, json_options);
+  if (status != Status::OK) {
+    LOG_WARN(
+        absl::StrCat("cannot parse logging plugin configuration JSON string ",
+                     configuration, ", ", status.message().ToString()));
+    return false;
+  }
+
+  if (config.logging_service().empty()) {
+    LOG_WARN(absl::StrCat(
+        "logging service has to be provided in the logging plugin config ",
+        configuration));
+    return false;
+  }
+
+  // Init gRPC call configuration.
+  logging_service_address_ = config.logging_service();
+  GrpcService grpc_service;
+  grpc_service.mutable_google_grpc()->set_target_uri(logging_service_address_);
+  grpc_service.SerializeToString(&grpc_service_);
+
+  // Init gRPC callbacks.
+  success_callback_ = [this](size_t) {
+    LOG_DEBUG("successfully sent loggin request");
+    in_flight_export_call_ -= 1;
+    if (in_flight_export_call_ < 0) {
+      LOG_WARN("in flight report call should not be negative");
+    }
+    if (in_flight_export_call_ <= 0 && is_on_done_) {
+      proxy_done();
+    }
+  };
+
+  failure_callback_ = [this](GrpcStatus status) {
+    LOG_WARN(absl::StrCat(
+        "Logging call error: ", std::to_string(static_cast<int>(status)),
+        getStatus().second->toString()));
+    in_flight_export_call_ -= 1;
+    if (in_flight_export_call_ < 0) {
+      LOG_WARN("in flight report call should not be negative");
+    }
+    if (in_flight_export_call_ <= 0 && is_on_done_) {
+      proxy_done();
+    }
+  };
+
+  // Start timer, which will trigger log report every 10s.
+  proxy_set_tick_period_milliseconds(10000 /* milliseconds */);
+
+  return true;
+}
+
+bool PluginRootContext::onDone() {
+  if (req_buffer_.empty() && in_flight_export_call_ == 0) {
+    return true;
+  }
+  // Flush out all log entries
+  flushLogBuffer();
+  sendLogRequest(/* ondone */ true);
+  return false;
```

Comment:
I think it's worth adding a comment explaining that `proxy_done` will signal completion when `false` is returned.